src/LuauExt.cpp

```diff
@@ -558,22 +581,58 @@ struct FindSymbolReferences : public Luau::AstVisitor
         return Luau::Symbol(name) == symbol;
     }
 
+    bool visitWriteExpr(Luau::AstExpr* expr) const
+    {
+        if (auto local = expr->as<Luau::AstExprLocal>(); local && visitLocal(local->local))
+            return true;
+        else if (auto global = expr->as<Luau::AstExprGlobal>(); global && visitGlobal(global->name))
+            return true;
+        return false;
+    }
+
     bool visit(Luau::AstStatLocalFunction* function) override
     {
         if (visitLocal(function->name))
         {
-            result.push_back(function->name->location);
+            addResult(function->name->location, lsp::DocumentHighlightKind::Write);
         }
         return true;
     }
 
+    bool visit(Luau::AstStatFunction* function) override
+    {
+        auto name = function->name;
+        if (visitWriteExpr(name))
+            addResult(name->location, lsp::DocumentHighlightKind::Write);
+        else
+            name->visit(this);
+        function->func->visit(this);
+        return false;
+    }
+
     bool visit(Luau::AstStatLocal* al) override
     {
         for (size_t i = 0; i < al->vars.size; ++i)
         {
-            if (visitLocal(al->vars.data[i]))
+            auto var = al->vars.data[i];
+            auto isSymbol = visitLocal(var);
+            if (symbol.local && i < al->values.size)
             {
-                result.push_back(al->vars.data[i]->location);
+                if (symbol.local->name == var->name && isRequire(al->values.data[i]))
+                {
+                    if (isSymbol)
```

Comment:
The check above only checks the name, not location or the other things, whereas, iirc, isSymbol is true when it truly is the symbol we're looking for, and false if it just has the same name. This way, the check in the else-if below prevents false positives when module imports get shadowed, as explained in the reply for that.

Reply:
Because the else-if below needs to be there if we want to keep module import / type prefix search completely sound. I could still rearrange the checks, getting rid of a level of indentation.